lib/ansible/plugins/action/fetch.py

```diff
@@ -161,29 +170,63 @@ def run(self, tmp=None, task_vars=None):
 
             dest = os.path.normpath(dest)
 
+            re_encrypt_needed = False
+            local_encrypted = False
+            b_file_path = to_bytes(dest, errors='surrogate_or_strict')
+            if self._loader.path_exists(b_file_path) and self._loader.is_file(b_file_path):
+                with open(to_bytes(dest), 'rb') as f:
+                    if self._loader._vault.is_encrypted(f.read(len(b_HEADER))):
+                        f.seek(0)
+                        data = f.read()
+                        if not self._loader._vault.secrets:
+                            raise AnsibleParserError("A vault password or secret must be specified to decrypt %s" % to_native(dest))
+
+                        data, vault_id_used, vault_secret_used = self._loader._vault.decrypt_and_get_vault_id(data, filename=dest)
+                        local_encrypted = True
+                        local_checksum = checksum_s(data)
+                        re_encrypt_needed = (vault_id_used != vault_id or vault_secret_used != vault_secret)
+
             # calculate checksum for the local file
-            local_checksum = checksum(dest)
+            if not local_encrypted:
+                local_checksum = checksum(dest)
 
-            if remote_checksum != local_checksum:
+            if remote_checksum != local_checksum or local_encrypted != encrypt or re_encrypt_needed:
                 # create the containing directories, if needed
                 makedirs_safe(os.path.dirname(dest))
 
                 # fetch the file and check for changes
                 if remote_data is None:
                     self._connection.fetch_file(source, dest)
-                else:
+                    if encrypt:
+                        with open(to_bytes(dest), "rb") as f:
+                            remote_data = f.read()
+                if remote_data is not None:
+                    if encrypt:
+                        remote_data = self._loader._vault.encrypt(remote_data, secret=vault_secret, vault_id=vault_id)
```

Comment:
This can cause an error by attempting to encrypt data that's already been encrypted with the same vault_id. To reproduce, I'm using this task:
```
  - name: Storing a file in an encrypted vault
    ansible.builtin.fetch:
      src: /tmp/encrypted
      dest: /tmp/dest
      flat: yes
      encrypt: yes
```
and `/tmp/encrypted` is encrypted with the default vault_id, and /tmp/dest exists and contains plaintext. I would expect the dest to be overwritten, but not for the source to be re-encrypted with the same vault_id, but I get:

```
fatal: [localhost]: FAILED! => {
    "msg": "input is already encrypted"
}
```

Reply:
You're probably right that the error is fine from a user-perspective. When I traced the code, I was interested in the codepath that inspects whether the source is encrypted and determines re_encrypt_needed, because I wasn't sure why the source would be encrypted. So I created a contrived example using an encrypted source to get to that codepath and see what happens if it theoretically shouldn't  need to be re-encrypted.

It sounds like this wasn't what you intended though - I think the logic for re_encrypt_needed can be removed in that case. This diff seems to pass with your tests:

```diff
diff --git a/lib/ansible/plugins/action/fetch.py b/lib/ansible/plugins/action/fetch.py
index af7de7be7a..73f1b33b8e 100644
--- a/lib/ansible/plugins/action/fetch.py
+++ b/lib/ansible/plugins/action/fetch.py
@@ -182,15 +182,15 @@ class ActionModule(ActionBase):
                             raise AnsibleParserError("A vault password or secret must be specified to decrypt %s" % to_native(dest))
 
                         data, vault_id_used, vault_secret_used = self._loader._vault.decrypt_and_get_vault_id(data, filename=dest)
-                        local_encrypted = True
-                        local_checksum = checksum_s(data)
-                        re_encrypt_needed = (vault_id_used != vault_id or vault_secret_used != vault_secret)
+                        if vault_id_used == vault_id or vault_secret_used == vault_secret:
+                            local_encrypted = True
+                            local_checksum = checksum_s(data)
 
             # calculate checksum for the local file
             if not local_encrypted:
                 local_checksum = checksum(dest)
 
-            if remote_checksum != local_checksum or local_encrypted != encrypt or re_encrypt_needed:
+            if remote_checksum != local_checksum:
                 # create the containing directories, if needed
                 makedirs_safe(os.path.dirname(dest))
```

That diff would also resolve https://github.com/ansible/ansible/pull/79864#discussion_r1098748255, since we'd no long decrypt random encrypted files with `encrypt: false`.

Reply:
You were right, my logic in this area was flawed. I made a change, please check if it is better now.
If you fetch encrypted file and specify encrypt=true, the file will still be double encrypted (user error I would say), but if you specify encrypt=no, then the file is transferred as is, not conversion done.